source/exe/signal_action.cc

```diff
@@ -7,6 +7,32 @@
 #include "common/common/assert.h"
 
 namespace Envoy {
+
+ABSL_CONST_INIT static absl::Mutex failure_mutex(absl::kConstInit);
+using FailureFunctionList = std::list<const CrashHandlerInterface*>;
+ABSL_CONST_INIT std::atomic<FailureFunctionList*> crash_handlers{nullptr};
```

Comment:
So, all the code here to me seems pretty problematic, but maybe there isn't a better way. We have mutexes, atomics with relaxed memory ordering, manual memory allocations, a singleton, etc. Oh my :)

One way we could be cleaner about this is to have a per-thread local optional signal handler for any signal. On a signal, the global handler will iterate over the list of threads and try and invoke the signal handler if registered.

Reply:
Yeah, I'd started by trying to implement this as a single handler in the server, and doing the plumbing through to snag dispatcher state, and it was messy enough I thought this was cleaner.  I think if we get to a point where more than one class wants to register signal handlers it'd might be cleaner to do a single entry point per thread but I think for now it's just moving the mess to a different location.